### PR TITLE
Start work on new Amount type

### LINKED ISSUE #126: Add an Amount type.
This should be a newtype wrapper around `i64` that enforces range constraints.  This means that arithmetic operations are fallible, so working with `Amount`s will require lifting into the `Option` monad, but by careful implementation of the `ops` traits, it should be possible to make this relatively ergonomic.

### PR DESCRIPTION
**Pull request by yaahc**
fixes https://github.com/ZcashFoundation/zebra/issues/126

### CODE REVIEW

**Comment by yaahc** [comment]
Stopping for the day but this is starting to take shape and is extremely cool

File: zebra-chain/src/types.rs
Comment (dconnolly) [inline zebra-chain/src/types.rs]: Cooool

File: zebra-chain/src/types.rs
Comment (dconnolly) [inline zebra-chain/src/types.rs]: I wonder if `#[derive(Arbitrary)]` works on `Amount`?

File: zebra-chain/src/types.rs
Comment (dconnolly) [inline zebra-chain/src/types.rs]: 🤓

**Review by dconnolly — DISMISSED** [review]
😎💃🎈

File: zebra-chain/src/types.rs
Comment (teor2345) [reply zebra-chain/src/types.rs]: I really like how this change helps us consistently check the `Amount` range in every calculation.

We should think about using a similar construct for other consensus-critical code where ranges are limited.

File: zebra-chain/src/types.rs
Comment (teor2345) [inline zebra-chain/src/types.rs]: Question: Should `arbitrary_with()` reject `Amount`s outside the valid range?

Should this code use `try_from()`, so all arbitrary amounts are valid?
If there is an issue here, why didn't our tests catch it?

File: zebra-chain/src/types.rs
Comment (teor2345) [inline zebra-chain/src/types.rs]: Issue: `Range` is lower-inclusive and upper-exclusive, but this code excludes both bounds.

Let's use `range.contains()` instead:
```suggestion
        if !range.contains(value) {
            Err("Amount's value is outside its valid range")
```

Comment (teor2345) [reply zebra-chain/src/types.rs]: Question: How can our tests catch issues like this?

Should the arbitrary tests have caught this issue?
Do we need more specific tests?

**Review by teor2345 — CHANGES_REQUESTED** [review]
Issue: There is an off-by-one error in the bounds.

I also have some non-blocking questions about testing.

Overall, I really like how this PR saves us from sprinkling range checks throughout our code.

File: zebra-chain/src/types.rs
Comment (yaahc) [reply zebra-chain/src/types.rs]: I don't think its necessary because this is generating the amounts _from_ the valid range, so its impossible for a subsequent `try_from` to fail by construction. We could use try_from to do the construction here but I don't see any proptest native way of dealing with fallible sources for `Strategy` impls, no `prop_try_map` or similar, so we would have to unwrap inside of the `arbitrary_with` impl. It should all optimize away but I don't think its necessary in the first place and strikes me as excessively defensive programming.

File: zebra-chain/src/types.rs
Comment (yaahc) [reply zebra-chain/src/types.rs]: Hmm, I wanted to try to get more specific error messages here but I can see the danger...

> Question: How can our tests catch issues like this?

:grimacing: not sure, being careful? Seems like we'd just have to remember to add a test for the edgecase to have caught this one.

File: zebra-chain/src/serialization.rs
Comment (hdevalence) [inline zebra-chain/src/serialization.rs]: These should be filled in before we merge this.

File: zebra-chain/src/transaction/transparent.rs
Comment (hdevalence) [inline zebra-chain/src/transaction/transparent.rs:86]: Is there some canonical description of how Bitcoin does this kind of checking? If we do this, we reject negative values during parsing.  Is that fine, if we need exact compatibility with Zcashd? I think it is probably fine but I'm not sure.

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: missing doc

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: I think this would be cleaner at the bottom of the file rather than in the middle

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: doc

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: doc

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: doc

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: doc

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: doc

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs:124]: If `C` is a marker type representing the bounds on `amount.0`, but we don't have any bounds on `C`, then we don't have any bounds on `amount.0`, right? In that case, how can we safely as-cast to `u64`?

Should this be `impl From<Amount<NonNegative>> for u64` or something similar?

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: I think it's preferable if the marker types `NegativeAllowed`, `NonNegative`, etc. are all empty enums rather than structs.  Because they are marker types, it makes no difference to the code whether they're enums or structs, but declaring them as enums means that they're listed in a separate section by rustdoc, which makes the API docs easier to read.

**Comment by hdevalence** [comment]
This looks great so far, I'm excited to see how all the `ops` impls work out.

One suggestion I have is that if you want to wait on writing docs until finishing the code, it's better to just comment out the `#![deny(missing_docs)] in a standalone commit you can drop later, rather than adding empty doc stubs, because the empty stubs are a lot harder to keep track of once they exist (since they work around the linter).

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: Is it okay if I leave this with the struct its impled on for now? If this merges before the arbitrary consolidation PR i'll move it as part of that PR and if the  other PR merges first I'll move it in this PR.

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs:124]: yes, good catch

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: Aah, I noticed you had them as enums but I didn't understand why, this makes sense :thumbsup:

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [reply zebra-chain/src/types/amount.rs]: Yeah sure sounds good

File: zebra-chain/src/types.rs
Comment (teor2345) [reply zebra-chain/src/types.rs]: Sounds good. I didn't know that input range was limited.

File: zebra-chain/src/types.rs
Comment (teor2345) [reply zebra-chain/src/types.rs]: Ideally, `Arbitrary` should test bounds as part of its early tests. But maybe it's not designed to do that.

File: zebra-chain/src/types/amount.rs
Comment (teor2345) [inline zebra-chain/src/types/amount.rs]: Nitpick: Let's document why conversion errors happen, and which errors are possible.

The current implementation only parses values from bytes. Because we parse bytes to `u64`, then convert to `i64`, we can get integer truncation errors if `value > i64::MAX`.

It took me a few minutes to work all that out, let's make it easier for the next person :-)

Here is a list of possible errors:
https://doc.rust-lang.org/std/num/enum.IntErrorKind.html

File: zebra-chain/src/types/amount.rs
Comment (teor2345) [inline zebra-chain/src/types/amount.rs]: Suggestion: Can we use a standard syntax to specify the range? (1/2)

The Rust range syntax:
```suggestion
/// Marker type for `Amount` that restricts the values to `-MAX_MONEY..=MAX_MONEY`
```

Or a mathematical range syntax:
```suggestion
/// Marker type for `Amount` that restricts the values to `-MAX_MONEY <= value <= +MAX_MONEY`
```

File: zebra-chain/src/types/amount.rs
Comment (teor2345) [inline zebra-chain/src/types/amount.rs]: Suggestion: Can we use a standard syntax to specify the range? (2/2)

The Rust range syntax:
```suggestion
/// Marker type for `Amount` that restricts the value to positive numbers `0..=MAX_MONEY`
```

Or a mathematical range syntax:
```suggestion
/// Marker type for `Amount` that restricts the value to positive numbers `0 <= value <= +MAX_MONEY`
```

**Review by teor2345 — DISMISSED** [review]
I am happy to merge this code, but I have a few suggestions for improving the comments.

File: zebra-chain/src/transaction/transparent.rs
Comment (daira) [reply zebra-chain/src/transaction/transparent.rs:86]: You'll have to reverse-engineer zcashd to get that information, I'm afraid. Please document it when you do!

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: I think you got `IntErrorKind` and `TryFromIntError` mixed up, the inner error here can actually only represent exactly one error which is a failure to convert a number that is out of range. This I feel makes your point even stronger, which is my way of saying :+1: .

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [inline zebra-chain/src/types/amount.rs:171]: @teor2345 I swapped these to `RangeInclusive` to hopefully prevent too many confusions over whether or not we should be including the bounds and because it let me remove a bunch of `+ 1`s. Also I swapped to using `displaydoc` for the `Display` impl because this seemed like a good candidate for deduping docs and error messages.

File: zebra-chain/src/transaction/serialize.rs
Comment (hdevalence) [inline zebra-chain/src/transaction/serialize.rs]: I'm a little confused since this `try_into` is the conversion into an `Amount`, right (?), but the error is being mapped to an `io::Error` even though there's a `SerializationError::Amount`.  Should it be using that error type instead, or is there a reason that it should be an `io::Error`?

File: zebra-chain/src/transaction/serialize.rs
Comment (hdevalence) [inline zebra-chain/src/transaction/serialize.rs]: is there a reason that `value_balance.into()` doesn't work here? (either way is fine, it doesn't matter, i'm just curious since i've run into issues with Into and references)

File: zebra-chain/src/transaction/serialize.rs
Comment (yaahc) [reply zebra-chain/src/transaction/serialize.rs]: iirc this function returns an `io::Error` we could map to a `SerializationError::Amount` first but then we'd still have to map to an `io::Error` unless we want to update the fn signature, let me take a look.

Comment (yaahc) [reply zebra-chain/src/transaction/serialize.rs]: nope i got it wrong, this should be a SerializationError, dont know how I got this one mixed up

File: zebra-chain/src/transaction/transparent.rs
Comment (hdevalence) [reply zebra-chain/src/transaction/transparent.rs:86]: Okay, we should probably split this question out into a separate issue then and unblock this PR.

Comment (hdevalence) [reply zebra-chain/src/transaction/transparent.rs:86]: #566

File: zebra-chain/src/transaction/serialize.rs
Comment (yaahc) [reply zebra-chain/src/transaction/serialize.rs]: it could be into, just needed to put parens around the deref

File: zebra-chain/src/types.rs
Comment (yaahc) [reply zebra-chain/src/types.rs]: I cant think of a way to do so but I'll try playing with it in the future

**Comment by hdevalence** [comment]
I think it would be good to have some of the `ops` impls in this PR, since those impls are pretty critical for this design and it would be good to know that they do actually work.

This would probably involve writing more tests, so the test module will get bigger and I think that it would make sense to move the test module to the bottom of the file as previously discussed.

File: zebra-chain/src/types/amount.rs
Comment (teor2345) [reply zebra-chain/src/types/amount.rs:171]: Thanks, I agree it's easier to understand when it is more explicit.

File: zebra-chain/src/types/amount.rs
Comment (teor2345) [inline zebra-chain/src/types/amount.rs:294]: Question: Is there a reason why we don't impl Add with option on both sides?

**Review by teor2345 — DISMISSED** [review]
Thanks, the calculations are going to be a lot easier with operators.

I had a minor question, and there's a merge conflict.

I would be happy to merge now, but let's check with @hdevalence.

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs:294]: because rust says it's a crime. Or more accurately it violates the orphan rules for trait impls which state you can't impl a foreign trait for a foreign type. In this case it's a little more complex because of the type parameter on the trait which makes covered types start to matter. for more details on the coherence rules check out https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: I think I'm missing something -- where does this do the validation?  Shouldn't there be a call to `AmountConstraint::validate` somewhere in here?

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs]: Hmm, if `C1` and `C2` are different constraints, how do we combine them into a `C1` for the `Output`?

Should this maybe be
```
impl<C: AmountConstraint> Add<Amount<C>> for Amount<C> {
    type Output = Option<Amount<C>>
    // ...
}
```

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs:143]: This impl requires a `Result` with an `Error` type. Together with the orphan impl problem we ran into, I think that this suggests that the `ops` impls should be `Result`-based rather than `Option`-based, for two reasons:

1. It allows writing, e.g., `a + b.try_into()` where `a: Amount`, `b: i32`;

2. It allows writing `(a + b)?` to cleanly handle failures.

Comment (hdevalence) [reply zebra-chain/src/types/amount.rs:143]: Otherwise we're stuck in a situation where some parts use Options and other parts use Results.

**Comment by hdevalence** [comment]
The coherence issues are unfortunate but I'm glad that we ran into them now rather than later.  I think that if we use `Result` rather than `Option`, it will be less of a problem because we can use `?` on sums and differences of `Amount`s.

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: Its being handled by the `this + rhs` expr which is forwarding to the Add impl without `Option`s on either side of the expr

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: Not sure I am undestanding what you mean by "how do we combine them", but this combines the constraints by doing the addition outside of the `Amount`. It would just unwrap the two Amounts regardless of their constraints, do the op, then pick one of the constraints as the output one and converts the result of the op into an Amount of that constraint, validating it along the way. I decided to default to the LHS as the constraint that gets used for the result because it seemed more intuitive. This lets you do things like `positive_amount + negative_amount` and be sure that the result is still positive.

If you think we should never add amounts with different constraints then we will probably need to add some conversion helpers to convert amounts of one constraint type into amounts of another constraint, however we probably can't do this with `From` because it will almost certainly overlap with `From<T> for T`.

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs:143]: big agree

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [inline zebra-chain/src/types/amount.rs:89]: oh this is really nice

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [reply zebra-chain/src/types/amount.rs]: Ah, that's quite nice

File: zebra-chain/src/types/amount.rs
Comment (hdevalence) [reply zebra-chain/src/types/amount.rs]: Yeah, what I mean is combining the constraints like you mention.  I think we should never add amounts with different constraints, unless we have a principled way to determine what the output constraint should be.  Otherwise, computing
```
positive_amt + negative_amt
```
and
```
negative_amt + positive_amt
```
would have different answers even though addition is supposed to be commutative. I think that this would be really surprising and error-prone.

We should definitely have a way to convert constraints as you mention.  If coherence prevents us from writing
```
impl<C1, C2> TryFrom<Amount<C1>> for Amount<C2>
```
then my next-best idea would be to write
```
impl<C1> Amount<C1> {
    pub fn constrain<C2>(&self) -> Result<Amount<C2>>
}
```
so that we can do
```
let positive_amount = negative_amount.constrain::<NonNegative>();
```

**Comment by hdevalence** [comment]
I understand the impl better now, my only remaining concern is about operations involving different constraints.

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: We might be able to add a 3rd parameter for the output and let inference take over.

edit: we cannot do this :(

Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: > If coherence prevents us from writing

Actually, thinking about it, coherence only disallows _generic_ impls for this, we can just manually specify the conversions between the two constraints we actually have I think.

Comment (yaahc) [reply zebra-chain/src/types/amount.rs]: I tried both using a manual `From`/`TryFrom` impls and ended up abandoning it because it wasn't able to infer the correct type for `into` through the generics on `Add`, the suggested `constrain` API works with inference quite nicely though so I went ahead and added it.

File: zebra-chain/src/types/amount.rs
Comment (yaahc) [inline zebra-chain/src/types/amount.rs:386]: @hdevalence new test showing the new constrain API for adding Amount's with different constraints

Comment (hdevalence) [reply zebra-chain/src/types/amount.rs:386]: this is really slick

**Review by hdevalence — APPROVED** [review]
LGTM

**Pull request merged**
